Feat/docs/audio visualizer tutorial (#275)
software-mansion-labs/react-native-audio-api: packages/audiodocs/docs/guides/see-your-sound.mdx

---
sidebar_position: 5
---

import InteractiveExample from '@site/src/components/InteractiveExample';

# See your sound

In this section, we explore the capabilities of the [`AnalyserNode`](/visualization/analyser-node) interface, focusing on techniques for extracting, processing, and visualizing audio data.

> Our exploration will lead us to create simple audio visualizer app.

Reviewer: Can you make it a bit more blunt (less whee-whoo GPT excitement 😅). I don't have anything against using GPT as a base for writing the docs, but usually it generates a bit of noise-words making text feeling a bit less imformative

Reviewer: Or to put this in other words: the simpler the language the better

> In this section, we explore the capabilities of the [`AnalyserNode`](/visualization/analyser-node) interface, focusing on techniques for extracting, processing, and visualizing audio data.
> Our exploration will lead us to create simple audio visualizer app.

Reviewer: ```suggestion
In this section, we explore the capabilities of the [`AnalyserNode`](/visualization/analyser-node) interface, focusing on techniques for extracting, processing, and visualizing audio data.
Our exploration will lead us to create simple audio visualizer app.

In this section, we will get familiar with capabilities of the [`AnalyserNode`](/visualization/analyser-node) interface, focusing on how to extract audio data in order to create a simple real-time visualization of the sounds. 
```

## Base application

> Building on what we have learned in the previous tutorials, we will begin by developing a straightforward application designed to play sound from a file. To begin, simply copy and paste the code provided below into your project.

Reviewer: ```suggestion
To kick-start things a bit, lets use code based on previous tutorials. It is a simple application that can load and play a sound from file. As previously if you would like to code along the tutorial, copy and paste the code provided below into your project.
```

      });

      setIsLoading(false);
    };

    fetchBuffer();

    return () => {
      audioContextRef.current?.close();
    };
  }, []);

  return (
    <View style={{ flex: 1}}>
      <View style={{ flex: 0.2 }} />
      <View
        style={{ flex: 0.5, justifyContent: 'center', alignItems: 'center' }}>
        {isLoading && <ActivityIndicator color="#FFFFFF" />}
        <View
          style={{
            justifyContent: 'center',
            flexDirection: 'row',
            marginTop: 16,
          }}>
        <Button
            onPress={handlePlayPause}
            title={isPlaying ? 'Pause' : 'Play'}
            disabled={!audioBufferRef.current}
            color={'#38acdd'}
          />
        </View>
      </View>
    </View>
  );
};

export default AudioVisualizer;
```

## Create an analyzer to capture and process audio data

To obtain frequency and time-domain data, we need to utilize the [`AnalyserNode`](/visualization/analyser-node).
It is an [`AudioNode`](/core/audio-node) that passes data unchanged from input to output while enabling the extraction of this data in two domains: time and frequency.

> We will use two specific [`AnalyserNode's`](/visualization/analyser-node) methods: [`getByteTimeDomainData`](/visualization/analyser-node#getbytetimedomaindata) and [`getByteFrequencyData`](/visualization/analyser-node#getbytefrequencydata).

Reviewer: As the link is provided in the first sentence, we can skip it here. (This is the wikipedia-way 🙂)
```suggestion
We will use two specific `AnalyserNode's` methods: [`getByteTimeDomainData`](/visualization/analyser-node#getbytetimedomaindata) and [`getByteFrequencyData`](/visualization/analyser-node#getbytefrequencydata).
```

Reviewer: Can you make the methods to be a short bullet list?

        FileSystem.documentDirectory + 'audio.mp3'
      ).then(({ uri }) => {
        return audioContextRef.current!.decodeAudioDataSource(uri);
      });

      setIsLoading(false);
    };

    fetchBuffer();

    return () => {
      audioContextRef.current?.close();
    };
  }, []);

  return (
    <View style={{ flex: 1}}>
      <View style={{ flex: 0.2 }} />
      <View
        style={{ flex: 0.5, justifyContent: 'center', alignItems: 'center' }}>
        {isLoading && <ActivityIndicator color="#FFFFFF" />}
        <View
          style={{
            justifyContent: 'center',
            flexDirection: 'row',
            marginTop: 16,
          }}>
        <Button
            onPress={handlePlayPause}
            title={isPlaying ? 'Pause' : 'Play'}
            disabled={!audioBufferRef.current}
            color={'#38acdd'}
          />
        </View>
      </View>
    </View>
  );
};

export default AudioVisualizer;
```

We utilize the [`requestAnimationFrame`](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestAnimationFrame) method to continuously fetch and update real-time audio visualization data.

## Visualize time-domain and frequency data

> We're ready to delve into visualizing our sound data using `react-native-skia`, exploring both time-domain and frequency-domain analyses.
> Before we begin, take a moment to review the differences between these two perspectives by visiting [Time domain vs Frequency domain](/visualization/analyser-node#time-domain-vs-frequency-domain).
> This background will enhance your understanding as we proceed with the visualizations.

Reviewer: ```suggestion
We're ready to delve into visualizing our sound data using `react-native-skia`, exploring both time-domain and frequency-domain analyses.
Before we begin, take a moment to review the differences between these two perspectives by visiting [Time domain vs Frequency domain](/visualization/analyser-node#time-domain-vs-frequency-domain).
This background will enhance your understanding as we proceed with the visualizations.

To render both the time as well as frequency domain visualizations, we will use our beloved graphic library - [`react-native-skia`](https://shopify.github.io/react-native-skia/). If you would like to know more what are time and frequency domains, have at look at [Time domain vs Frequency domain](/visualization/analyser-node#time-domain-vs-frequency-domain) section of the AnalyserNode documentation, which explains those terms in details, but otherwise here is the code:
```
